nixos/modules/services/networking/prosody.nix

```diff
@@ -1,887 +1,675 @@
+      example = literalExpression ''
+        {
+          "example.net" = {
+            useACMEHost = "example.net";
+            settings = {
+              admins = [ "admin1@example.net" "admin2@example.net" ];
+              c2s_require_encryption = true;
+              modules_enabled = [
+                "announce"
+              ];
+            };
+          };
+        }
+      '';
+    };
 
-      authentication = mkOption {
-        type = types.enum [ "internal_plain" "internal_hashed" "cyrus" "anonymous" ];
-        default = "internal_hashed";
-        example = "internal_plain";
-        description = lib.mdDoc "Authentication mechanism used for logins.";
-      };
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
+      description = mdDoc ''
+        Additional prosody configuration.
+      '';
+    };
 
-      extraConfig = mkOption {
-        type = types.lines;
-        default = "";
-        description = lib.mdDoc "Additional prosody configuration";
-      };
+    # options to preserve compatibility with this module pre nixos 23.11
 
+    modules = mkOption {
+      type = types.attrsOf types.bool;
+      default = { };
+      description = mdDoc ''
+      '';
     };
   };
 
-
-  ###### implementation
-
   config = mkIf cfg.enable {
 
-    assertions = let
-      genericErrMsg = ''
-
-          Having a server not XEP-0423-compliant might make your XMPP
-          experience terrible. See the NixOS manual for further
-          information.
-
-          If you know what you're doing, you can disable this warning by
-          setting config.services.prosody.xmppComplianceSuite to false.
-      '';
-      errors = [
-        { assertion = (builtins.length cfg.muc > 0) || !cfg.xmppComplianceSuite;
-          message = ''
-            You need to setup at least a MUC domain to comply with
-            XEP-0423.
-          '' + genericErrMsg;}
-        { assertion = cfg.uploadHttp != null || !cfg.xmppComplianceSuite;
-          message = ''
-            You need to setup the uploadHttp module through
-            config.services.prosody.uploadHttp to comply with
-            XEP-0423.
-          '' + genericErrMsg;}
-      ];
-    in errors;
-
-    environment.systemPackages = [ cfg.package ];
-
-    environment.etc."prosody/prosody.cfg.lua".text =
+    assertions =
       let
-        httpDiscoItems = optionals (cfg.uploadHttp != null)
-            [{ url = cfg.uploadHttp.domain; description = "HTTP upload endpoint";}];
-        mucDiscoItems = builtins.foldl'
-            (acc: muc: [{ url = muc.domain; description = "${muc.domain} MUC endpoint";}] ++ acc)
-            []
-            cfg.muc;
-        discoItems = cfg.disco_items ++ httpDiscoItems ++ mucDiscoItems;
-      in ''
-
-      pidfile = "/run/prosody/prosody.pid"
-
-      log = "*syslog"
-
-      data_path = "${cfg.dataDir}"
-      plugin_paths = {
-        ${lib.concatStringsSep ", " (map (n: "\"${n}\"") cfg.extraPluginPaths) }
```

Comment:
This used to define an empty block if no extraPluginPaths were defined -- not sure if this inhibits any prosody default or not, need to check later™

(Same for `s2s_insecure_domains` and `s2s_secure_domains` but these don't sound like they'd have default values if unset; plugin paths is more dubious to me)

Reply:
`plugin_paths` is marked as `--plugin_paths = {}` in the upstream example file which leads me to believe this is the default and fine

i'll use the same idea that lists are additive and assume `s2s_insecure_domains` and `s2s_secure_domains` are fine...

Reply:
Yes this makes sense with the other list being additive for virtualhost -- thanks for checking.